design/proposals/ambient-namespaces.md

```diff
@@ -0,0 +1,52 @@
+# Ambient Namespaces
+
+Status: Pending
+
+Version: Alpha
+
+## Abstract
+
+We want to introduce functionality to enable the use of environment metadata (ex: cluster namespace, Server URI) in prototype and component manifest files.
```

Comment:
The idea that we would expose cluster name or server URI is somewhat controversial. (I am personally for it, but I think everyone agrees, as you mention below, having people write stuff like `if (env.name) then [...]` is considered an anti-pattern, which is why other people on the team feel strongly we should not do this.)

So: did you mean to expose all of this metadata, or did you only want the namespace? Either one is ok with me, but I think we should beef up the motivation for the other two if we're going to go that route.

FWIW, personally I think there is a good case to be made that namespace is actually special, since it is actually part of the app code (_i.e._, we specify it in the YAML), while name and server URI are not (_i.e._, they do not show up in the YAML).

Reply:
The intention was to expose all this metadata, but that was with the assumption that future `ksonnet-lib` APIs may take a server address. i.e. That it may be part of the future app code.

If this is a valid assumption, I'll beef up the motivation section with this point.

Reply:
Oh, interesting, I actually can't think of a time we'd want to have server URI exposed as part of ksonnet-lib. What sort of scenarios were you thinking of?